add Q2 + modify Q1

nuwanda57 opened the pull request.

nuwanda57 committed:
Q2 solution

nuwanda57 committed:
improved Q1

improve style

ssancho2 commented on Q2.cpp:
```diff
@@ -0,0 +1,83 @@
+
+Node* CreateSinglyLinkedList() {
+    Node *Head = new Node;
+    std::cin >> Head->value;
```
With all the inputs from the user it'll be nice to have some kind of prompt on what to enter when.

ssancho2 replied:
Also, instead (or in addition to) the user entering the data, it may be good to have some code that creates test cases and compares the results against expected values, that way you don't have to manually enter them each time you want to check your code. A common pattern is:
void doAllTests() {
    doOneTest(input, expected output);
    ....
}

ssancho2 commented on Q2.cpp:
```diff
@@ -0,0 +1,83 @@
+
+
+struct Node {
+    char value;
```
Not a big deal, and not part of the problem, but it's nicer to use a templated type:
template <typename T>
struct Node {
     T value;
     Node* next;
 };

ssancho2 commented on Q2.cpp:
```diff
@@ -0,0 +1,83 @@
+    return (Head);
+};
+
+Node* KthToLast(Node* Head, int K, bool &problem) {
```
No need for a bool parameter, you can return NULL when the input is invalid.
Also, reference parameters are usually discouraged, because it's not clear when you call the function that the parameter may be modified, we usually use pointers for that.
Compare:
Node* Result = KthToLast(Head, K, problem);
With
Node* Result = KthToLast(Head, K, &problem);
In the second is very clear that 'problem' may be modified.

ssancho2 replied:
What happens if Head itself is NULL, or k is negative?

ssancho2 commented on Q2.cpp:
```diff
@@ -0,0 +1,83 @@
+        }
+    }
+    // Now we use another pointer with the same speed,
+    // that stars from the Head
```
stars -> starts

ssancho2 commented on Q2.cpp:
```diff
@@ -0,0 +1,83 @@
+        Elem = Elem->next;
+        FutureResult = FutureResult->next;
+    }
+    return (FutureResult);
```
Why the parenthesis?

ssancho2 reviewed (approved):
Looks good, some minor comments.

nuwanda57 committed:
Q2_changed1

use templates
add prompt on what to enter
KthToLast function returns NULL in case of a problem
delete problem variable
k < 0: IMPOSSIBLE INPUT
empty list: IMPOSSIBLE INPUT
stars -> starts
delete  parenthesis

And also I have a question:
I want to give user an opportunity to enter a typename. But I don't know a better way how to resolve it than to compare his answer to all possible types. Is there a better way?

julia57 commented:
change Q2

julia57 commented:
Hi!Thank you for the reviews! I've made some changes and commited, but I'm not sure if you can see these changes, because I've commited to the same branch, can you? Many thanks,Julia  18.05.2017, 21:31, "Sergio Sancho" <notifications@github.com>:@ssancho2 approved this pull request.Looks good, some minor comments.In Q2.cpp:> +//      if K is bigger than the size of the list the answer will be IMPOSSIBLE INPUT
+//      KthToLast function contains a solution
+//
+
+#include <iostream>
+#include <vector>
+
+
+struct Node {
+    char value;
+    Node* next;
+};
+
+Node* CreateSinglyLinkedList() {
+    Node *Head = new Node;
+    std::cin >> Head->value;
With all the inputs from the user it'll be nice to have some kind of prompt on what to enter when.In Q2.cpp:> +//
+//  problem specification:
+//      Implement an algorithm to find the kth to last element of a singly linked list
+//
+//  comments:
+//      list length is unknown
+//      if K is bigger than the size of the list the answer will be IMPOSSIBLE INPUT
+//      KthToLast function contains a solution
+//
+
+#include <iostream>
+#include <vector>
+
+
+struct Node {
+    char value;
Not a big deal, and not part of the problem, but it's nicer to use a templated type:templatestruct Node {T value;Node* next;};In Q2.cpp:> +
+Node* CreateSinglyLinkedList() {
+    Node *Head = new Node;
+    std::cin >> Head->value;
+    Node *elem1 = Head;
+    Node *elem2 = new Node;
+    while (std::cin >> elem2->value) {
+        elem1->next = elem2;
+        elem1 = elem2;
+        elem2 = new Node;
+    }
+    elem1->next = NULL;
+    return (Head);
+};
+
+Node* KthToLast(Node* Head, int K, bool &problem) {
No need for a bool parameter, you can return NULL when the input is invalid.Also, reference parameters are usually discouraged, because it's not clear when you call the function that the parameter may be modified, we usually use pointers for that.Compare:Node* Result = KthToLast(Head, K, problem);WithNode* Result = KthToLast(Head, K, &problem);In the second is very clear that 'problem' may be modified.In Q2.cpp:> +};
+
+Node* KthToLast(Node* Head, int K, bool &problem) {
+    Node* Elem = Head;
+    // pass through the list till K nodes are passed
+    for (int i = 0; i < K; ++i) {
+        if (Elem->next != NULL)
+            Elem = Elem->next;
+        else { // K is bigger than the list length
+            problem = true;
+            std::cout << "IMPOSSIBLE INPUT" << std::endl;
+            return Elem;
+        }
+    }
+    // Now we use another pointer with the same speed,
+    // that stars from the Head
stars -> startsIn Q2.cpp:> +        else { // K is bigger than the list length
+            problem = true;
+            std::cout << "IMPOSSIBLE INPUT" << std::endl;
+            return Elem;
+        }
+    }
+    // Now we use another pointer with the same speed,
+    // that stars from the Head
+    Node *FutureResult = Head;
+    // When the first pointer reaches the end of the list
+    // The second pointer points at the answer
+    while (Elem->next != NULL){
+        Elem = Elem->next;
+        FutureResult = FutureResult->next;
+    }
+    return (FutureResult);
Why the parenthesis?In Q2.cpp:> +
+Node* CreateSinglyLinkedList() {
+    Node *Head = new Node;
+    std::cin >> Head->value;
+    Node *elem1 = Head;
+    Node *elem2 = new Node;
+    while (std::cin >> elem2->value) {
+        elem1->next = elem2;
+        elem1 = elem2;
+        elem2 = new Node;
+    }
+    elem1->next = NULL;
+    return (Head);
+};
+
+Node* KthToLast(Node* Head, int K, bool &problem) {
What happens if Head itself is NULL, or k is negative?—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.   

ssancho2 commented on Q2.cpp:
```diff
@@ -7,77 +7,92 @@
-    Node* Elem = Head;
+template <typename T>
+Node<T>* KthToLast(Node<T>* Head, int K) {
+    if (K < 0 || Head->value == NULL)
```
Head == NULL
We don't care that much about the values :)

ssancho2 commented on Q2.cpp:
```diff
@@ -7,77 +7,92 @@
 
-int main() {
+template <typename T>
+void MainFunction() {
```
Templating this function may be pushing it too far (but if you do, you can define K to be of type T).

ssancho2 reviewed (approved):
Regarding entering a typename, I don't think there is an easy way to do stuff like that in c++. To implement something like that you can look into the 'abstract factory' design patter, or a class registry.

nuwanda57 committed:
Update Q2 version2

delete function MainFunction
(Head->value == NULL) --> (Head == NULL)

And I have one more question (not about the program):
I use construction "while (cin >> x) {}" and I don't know how to terminate input in my code editor (CLion), so I needed to test my program in another editor. Maybe you know how I can do it. I tried ctrl+d and many other variants but it did't work.
Thanks!

ssancho2 reviewed (commented):
You may try changing it to :
while (!(cin >> x).eof()) { }
cin >> x returns cin, so you can chain several '>>' calls, implicit conversion from istream to bool may not be reliable.
